[Jeff Sieu] iP

jeffsieu opened the pull request:
# Changes

- Adds task list functionality
- Tasks _may_ have deadlines or dates
- Tasks are saved locally to a text file named `duke-task-list.txt`.

# Progress

- [x] Add text-based command functionality
- [x] Add save feature
- [ ] Add GUI

# Code structure

The code is split into packages, namely:
1. `duke`
2. `duke.command`
3. `duke.exception`
4. `duke.task`

## DukeCommand implementation

The [DukeCommand](https://github.com/nus-cs2103-AY2122S1/ip/blob/31d7d3184f03544f347477375b4d419d984647dd/src/main/java/duke/command/DukeCommand.java) is implemented as an enum, as there is only a pre-defined set of commands in Duke at any time.

> A Java enum allows for more than just fixed values; it allows a fixed set of object instances, which can each contain their own methods and variables. A constructor can also be defined.

The above feature of Java enums allows the commands in Duke to be pre-defined and _prevents the wrongful instantiation_ of commands elsewhere in the code.

As a result, all commands are encapsulated and declared within the `DukeCommand.java` file. 😄 

Each package contains files relevant to that aspect of Duke.

This can be seen in the following code snippet showing `HELP`, the help command:

```java
HELP("help",
            "Display all commands, or detailed info of given command",
            new DukeCommandConfig(
                    new DukeCommandArgument("command_name", "The name of the command to view help for",
                            DukeCommandArgumentType.OPTIONAL), Map.of()),
            (TaskList taskList, Ui ui, Storage storage, String arg, Map<String, String> namedArgs) -> {
                if (arg.isEmpty()) {
                    ui.outputLine("ALL COMMANDS:");
                    Arrays.stream(DukeCommand.values())
                            .sorted(Comparator.comparing(DukeCommand::getName))
                            .map(DukeCommand::toString)
                            .forEach(ui::outputLine);
                } else {
                    Optional<DukeCommand> command = getClosestMatch(arg);
                    if (command.isEmpty()) {
                        throw new InvalidCommandException(String.format("No command named \"%s\".", arg));
                    } else {
                        ui.outputLine(command.get().toDetailedString());
                    }
                }
            }),
```

Commit:
Add Gradle support

jeffsieu committed:
Skeletal version: Greet, Echo, Exit

- Duke greets the user
- All commands are echoed
- "bye" causes the Duke to exit

jeffsieu committed:
Add task addition and listing

Duke now allows the user to list tasks with the
list command, and add tasks with any command other
than list or bye.

jeffsieu committed:
Add mark as done, represent commands as enums

jeffsieu committed:
Update commands to allow named arguments

jeffsieu committed:
Add event and deadline tasks, display task by type

jeffsieu committed:
Add basic testing

jeffsieu committed:
Add error handling for argument mismatch

jeffsieu committed:
Finish error handling

jeffsieu committed:
Add task deletion

jeffsieu committed:
Add save feature - Level 7

jeffsieu committed:
Add date recognition - Level 8

jeffsieu committed:
Merge branch 'branch-Level-7'

jeffsieu committed:
Merge branch 'branch-Level-8'

jeffsieu committed:
Fix errors from merging branches

jeffsieu committed:
Refactor storage functions to Storage class

jeffsieu committed:
Fix code warnings

jeffsieu committed:
Refactor list of tasks to TaskList

jeffsieu committed:
Refactor interactions to Ui

- Rename DukeCommandFormatter to Ui
- Move all console printing related operations
  to Ui

jeffsieu committed:
Refactor parsing logic to Parser

jeffsieu committed:
Clean up code

jeffsieu committed:
Divide classes into packages

jeffsieu committed:
Add JUnit tests

jeffsieu committed:
Refactor program exit logic

jeffsieu committed:
Add JavaDoc comments

jeffsieu committed:
Follow coding standard

jeffsieu committed:
Add find command

jeffsieu committed:
Merge branch 'branch-A-CodingStandard'

jeffsieu committed:
Merge branch 'branch-Level-9'

jeffsieu committed:
Add META-INF folder

jeffsieu committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

jeffsieu committed:
Add checkstyle

jackgugz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+    public Duke() {
+        this.ui = new Ui(System.in, System.out);
+        this.storage = new Storage(SAVE_FILE_LOCATION);
+        TaskList taskList;
```
Do you think we can remove this line?

jackgugz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+        ui.printWelcomeMessage();
+        ui.printHelp();
+        while (ui.shouldContinue()) {
+            String commandStr = ui.nextCommand();
```
I can understand this but I am not sure if we should use standard word String or call it `commandInput` the Str instead as it will be hard to pronounce this way. 
```suggestion
            String commandInput = ui.nextCommand();
```

jackgugz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,37 @@
+        Optional<DukeCommand> dukeCommand = DukeCommand.getClosestMatch(command);
+        if (dukeCommand.isPresent()) {
+            DukeCommand actualCommand = dukeCommand.get();
+            String arguments = command.substring(actualCommand.getName().length());
```
Maybe we can specify here what arguments it is?

jackgugz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    public DukeCommandWithArgs parse(String command) {
+        // Get the longest duke command that matches to command
+        Optional<DukeCommand> dukeCommand = DukeCommand.getClosestMatch(command);
+        if (dukeCommand.isPresent()) {
+            DukeCommand actualCommand = dukeCommand.get();
+            String arguments = command.substring(actualCommand.getName().length());
+            String[] tokens = arguments.split("/");
+            String positionalArg = tokens[0].trim();
+            Map<String, String> namedArgs = new HashMap<>();
+            for (int i = 1; i < tokens.length; i++) {
+                String[] namedArg = tokens[i].trim().split(" ", 2);
+                namedArgs.put(namedArg[0].trim(), namedArg[1].trim());
+            }
```
Sorry Jeff but I am a bit confused over here would you explain maybe in comment form on what is happening here?

jackgugz commented on src/main/java/duke/command/DukeCommand.java:
```diff
@@ -0,0 +1,248 @@
+package duke.command;
+
+import duke.Storage;
+import duke.Ui;
+import duke.exception.InvalidCommandException;
+import duke.task.DukeDeadlineTask;
+import duke.task.DukeEvent;
+import duke.task.DukeSimpleTask;
+import duke.task.DukeTask;
+import duke.task.TaskList;
```
Maybe we can improve the ordering here?

jackgugz commented on src/main/java/duke/command/DukeCommand.java:
```diff
@@ -0,0 +1,248 @@
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public enum DukeCommand implements DukeCommandAction {
```
I believe I desperately need some comments here over what is happening. Would you kindly help me with that by adding in some comments?

tanjoewel replied:
I agree with this review I also need some comments to understand what is happening

jackgugz commented on src/main/java/duke/command/DukeCommand.java:
```diff
@@ -0,0 +1,248 @@
+     * Returns the name of the command.
+     * @return the name of the command
+     */
+    public String getName() {
```
May I know what name is it referring to? The name of the command or do you mean the type of the command?

charliemoweng replied:
nice to see that there has been javadoc comment added to explain it better.

jackgugz reviewed (commented):
Overall almost all the namings are done in line with the guidelines given to us. Super impressed with the readability and neatness of the code. Just that due to the complexity of command-processing classes, it may be best to be commented with more explanation on how the entire process works to a rockie outsider coder like me. :D 
Great Job!

tanjoewel commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,67 @@
+        this.parser = new Parser();
+    }
+
+    /**
+     * Runs Duke.
+     * @param args The arguments passed in from the command line
+     */
```
Should there be an empty line between the description and parameter section?
It is mentioned in the coding standard, and I noticed this multiple times

tanjoewel reviewed (commented):
LGTM! I couldn't find any other coding standard violation other than the JavaDoc one. Overall the code is very well written and easy to navigate, but I agree with @jackgugz that the use of Enums, Optionals and Streams can be difficult to understand for inexperienced coders so would be good to have comments describing whats going on.
Good job!

jeffsieu committed:
Add GUI

jeffsieu committed:
Improve DukeCommand documentation

jeffsieu committed:
Add delete button

charliemoweng commented on src/main/java/duke/ActionButtonTableCell.java:
```diff
@@ -0,0 +1,43 @@
+import javafx.scene.control.TableColumn;
+import javafx.util.Callback;
+
+public class ActionButtonTableCell<S> extends TableCell<S, Button> {
```
maybe can add in some Javadoc?

charliemoweng commented on src/main/java/duke/Cli.java:
```diff
@@ -0,0 +1,125 @@
+
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.List;
```
maybe can separate them by one line? so the import statements are grouped into groups.

charliemoweng commented on src/main/java/duke/Gui.java:
```diff
@@ -0,0 +1,234 @@
+
+import java.time.LocalDate;
+
+import duke.exception.DukeStorageException;
```
that's a lot of import statements! maybe it will look nicer if you group them accordingly?

charliemoweng commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,4 @@
+package duke;
+
+public interface Ui {
```
is this still used?

charliemoweng reviewed (commented):
No major issues to me. Only styles issues here and there. I feel that I can learn a lot from you. Keep up the good work!

jeffsieu committed:
Add assertions

jeffsieu committed:
Improve code quality

Code: refactor to match code quality standards

There are some issues with variable names and some syntatic
constructs that are less ideal according to the code quality guide.

Refactoring the code to match the standards on the code quality guide
will make it easier to understand and maintain.

jeffsieu committed:
Merge pull request #1 from jeffsieu/branch-A-CodeQuality

Improve code quality

jeffsieu committed:
Merge branch 'master' of https://github.com/jeffsieu/ip

jeffsieu committed:
Merge branch 'master' into branch-A-Assertions

jeffsieu committed:
Merge pull request #2 from jeffsieu/branch-A-Assertions

Add assertions

jeffsieu committed:
Add User Guide

jeffsieu committed:
Add cli flag as program argument

jeffsieu committed:
Fix build.gradle

jeffsieu committed:
Fix docs/README.md

jeffsieu committed:
Fix smoke test error

jeffsieu committed:
Update README.md with clearer instructions

jeffsieu committed:
Update README.md

jeffsieu committed:
Update README.md
